.github/workflows/test.yaml

```diff
@@ -513,16 +512,47 @@ jobs:
       run: |
         ls -R dist/
 
+    - uses: actions/upload-artifact@v4
+      with:
+        name: dist
+        path: dist
+
     - name: Check matched tag version and branch version - on tag
       if: startsWith(github.ref, 'refs/tags/twisted-')
       run: python admin/check_tag_version_match.py "${{ github.ref }}"
 
-    - name: Publish to PyPI - on tag
-      if: startsWith(github.ref, 'refs/tags/twisted-')
-      uses: pypa/gh-action-pypi-publish@v1.5.1
+  # Used exclusively for various release publishing.
+  # This is only executed on tags, but just does the publish to exercise the release as much
+  # as possible and reduce the possibility of finding bugs in the release
+  # process late in the release cycle,
+  release-publish:
+    needs:
+      release-build
+    if: startsWith(github.ref, 'refs/tags/twisted-')
+    environment:
+      release
+    permissions:
+      # IMPORTANT: this permission is mandatory for trusted publishing
+      id-token: write
+    name: publish on twisted-* tag
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/download-artifact@v4
       with:
-        password: ${{ secrets.PYPI_UPLOAD_TOKEN }}
+        name: dist
+        path: dist
+
+    - name: Publish to PyPI - on tag
```

Comment:
can we skip this step, instead of the whole job?

In this way, we should no longer need ` release-publish-success-or-skipped` ?

Reply:
No, because the environment is not available to branches

Reply:
I missed the GitHub environment part.

Do we need an environment ?

I feel that we are adding a lot of complexity and will make it harder for future maintainer to know what is going on there.

I think that this is the first place where we are introducing GItHub Action Environments in our CI.

If this is the first place, we need to add some documentation about how we plan to use them and how to configure them

I see a setting at https://github.com/twisted/twisted/settings/environments

Reply:
An environment is strongly recommended, and helps prevent accidents

Reply:
I know that in big companies, where you have a lot of developers that have no idea what they are doing, environments are of great help.

but I think that for the Twisted team, we can live with a bit of risk :)

I am not -1 in terms of environments, but we need to document them and update the release process, so that the person who will try to do a release will know which "wheel" are spinning during the release process

Reply:
Happy to add a document for this

Reply:
I've added some notes to the release process docs